### PR TITLE
implement contract events

### PR DESCRIPTION
**Pull request by thiolliere**
implement `ext_send_event(data: Vec<u8>)` that register `Contract(AccountId of contract, data)`

fee is computed with new parameter : event_data_per_byte_cost

### CODE REVIEW

**Comment by pepyakin** [comment]
I am not sure if it's worth updating COMPLEXITY.md since now it got a little bit outdated and at the same time we are expecting big changes for state rent.

**Comment by gavofyork** [comment]
Yeah - please keep it up to date will all changes in code as they happen.

**Comment by gavofyork** [comment]
Both TODOs should be done before merge (update complexity.md and introduce a base fee)

**Comment by pepyakin** [comment]
Regarding base fees: As @thiolliere noted, there are no other instances of "base fees" except for call and create/instantiate. This is because having a base fee is an open question...

This doesn't necessary mean that this function presents an attack vector though. We charge gas for a call, for reading memory for N bytes and for putting an event which is N bytes, which sounds good enough. We still might want to add some fee to account that this is actually a call to the runtime which crosses the sandbox boundary which might be a somewhat expensive. However, at the moment I am slightly more inclined to take this fee before calling to such a function, i.e. at the gas metering level.

However, I think this might be out of scope for this PR.

Ah yeah complexity.md might indeed benefit from detailing the deposit event point.

**Comment by thiolliere** [comment]
complexity details improved, base fee is just a matter of few lines though

File: srml/contract/src/wasm/mod.rs [hunk 1]
```diff
@@ -276,6 +276,9 @@ mod tests {
 		fn random_seed(&self) -> &H256{
 			&self.random_seed
 		}
+
+		fn deposit_event(&mut self, _data: Vec<u8>) {
```
Comment (tomusdrw) [inline srml/contract/src/wasm/mod.rs]: Maybe make it `unimplemented!()` to easily catch potential issues in the test code.

Comment (thiolliere) [reply srml/contract/src/wasm/mod.rs]: yes I finally implement a test that checks charges

**Comment by gavofyork** [comment]
Why is the base cost reverted!?

Adding an event is a significant cost in and of itself. adding a 1K event is not 1000x more resource intensive than adding a 1 byte event.

**Comment by pepyakin** [comment]
@gavofyork yeah, I asked @thiolliere to do that. I think this is not principal for me and can be re-reverted, but I'd prefer to addess all these issues (you're right that other externalities lack of proper base fees) all at once in some general fashion and I thought that it would be better if we did that in a separate issue and didn't have any ad-hoc solutions built in already.

**Comment by gavofyork** [comment]
It's quite possible that not all other externalities will need base costs and I don't see it as a single feature that deserves its own PR. This PR should be about introducing a *non-buggy* contract events. For it to be non-buggy it *needs a base cost*. If other externalities also need base costs, then it's a different, bug-fix, PR.

**Pull request merged**